src/iqm/cirq_iqm/serialize.py

```diff
@@ -16,11 +16,139 @@
+from iqm.cirq_iqm.iqm_gates import IQMMoveGate
+from iqm.iqm_client import Instruction
+
+# Mapping from IQM operation names to cirq operations
+_IQM_CIRQ_OP_MAP: dict[str, tuple[type[Gate], ...]] = {
+    # XPow and YPow kept for convenience, Cirq does not know how to decompose them into PhasedX
+    # so we would have to add those rules...
+    'prx': (PhasedXPowGate, XPowGate, YPowGate),
+    'cz': (CZPowGate,),
+    'move': (IQMMoveGate,),
+    'measure': (MeasurementGate,),
+    'cc_prx': (PhasedXPowGate, XPowGate, YPowGate),
+}
+
+
+def instruction_to_operation(instr: Instruction) -> Operation:
+    """Convert an IQM instruction to a Cirq Operation.
+
+    Args:
+        instr: the IQM instruction
+
+    Returns:
+        Operation: the converted operation
+
+    Raises:
+        OperationNotSupportedError When the circuit contains an unsupported operation.
+
+    """
+    if instr.name not in _IQM_CIRQ_OP_MAP:
+        raise OperationNotSupportedError(f'Operation {instr.name} not supported.')
+
+    gate_type = _IQM_CIRQ_OP_MAP[instr.name][0]
+    args_map = {'angle_t': 'exponent', 'phase_t': 'phase_exponent'}
+    args = {args_map[k]: v * 2 for k, v in instr.args.items() if k in args_map}
+    if 'key' in instr.args.keys():
+        args['key'] = instr.args['key']
+        args['num_qubits'] = len(instr.qubits)
+    qubits = [NamedQid(qubit, dimension=2) for qubit in instr.qubits]
+    if instr.name == 'cc_prx':
+        return gate_type(**args)(*qubits).with_classical_controls(instr.args['feedback_key'])
+    return gate_type(**args)(*qubits)
+
+
+class OperationNotSupportedError(RuntimeError):
+    """Raised when a given operation is not supported by the IQM server."""
+
+
+def map_operation(operation: Operation) -> Instruction:
+    """Map a Cirq Operation to the IQM data transfer format.
+
+    Assumes the circuit has been transpiled so that it only contains operations natively supported by the
+    given IQM quantum architecture.
+
+    Args:
+        operation: a Cirq Operation
+
+    Returns:
+        Instruction: the converted operation
+
+    Raises:
+        OperationNotSupportedError When the circuit contains an unsupported operation.
+
+    """
+    qubits = [qubit.name if isinstance(qubit, NamedQid) else str(qubit) for qubit in operation.qubits]
+    if isinstance(operation.gate, (PhasedXPowGate, XPowGate, YPowGate)):
+        return Instruction(
+            name='prx',
+            qubits=tuple(qubits),
+            args={'angle_t': operation.gate.exponent / 2, 'phase_t': operation.gate.phase_exponent / 2},
+        )
+    if isinstance(operation.gate, MeasurementGate):
+        if any(operation.gate.full_invert_mask()):
+            raise OperationNotSupportedError('Invert mask not supported')
+
+        return Instruction(
+            name='measure',
+            qubits=tuple(qubits),
+            args={'key': operation.gate.key},
```

Comment:
```suggestion
            args={'key': operation.gate.key, 'feedback_key': operation.gate.key},
```
As per @smite 's earlier comments. The 'key' is used for returning the measurement in the results and 'feedback_key' is used for the cc_prx.